plugins/pulp_rpm/plugins/importers/yum/purge.py

```diff
@@ -286,3 +289,209 @@ def remove_unit_duplicate_nevra(unit, repo):
+            # q objects don't deal with order_by, so they can't be used with repo_controller funcs
+            # disassociate_units only uses the unit_id, so limit the resultset to only that field
+            rcus = model.RepositoryContentUnit.objects.filter(
+                repo_id=repo_id, unit_id__in=unit_ids).order_by('-updated').only('unit_id')
+
+            # 0 or 1 packages from the duplicate nevra search match this repo means no duplicates
+            if rcus.count() < 2:
+                continue
+
+            repo = model.Repository.objects.get(repo_id=repo_id)
+
+            # Since the repo_units queryset is ordered by the updated field (descending), the
+            # first repo content unit is the latest. All other RCUs should be disassociated
+            duplicate_units = (unit_type(id=rcu.unit_id) for rcu in rcus[1:])
+            repo_controller.disassociate_units(repo, duplicate_units)
+
+
+def _duplicate_key_nevra_fields(unit):
+    """strip out fields not related to nevra from a nevra-supporting unit key
+
+    :param unit: unit with NEVRA fields to be filtered
+    :type unit: ContentUnit
+    """
+    # consider duplicates to be units with the same unit key when the checksum is ignored.
+    # normally that means NEVRA, but with DRPMs, for example, filename is used instead of name
+    fields = list(unit.NAMED_TUPLE._fields)
+    fields.remove('checksum')
+    fields.remove('checksumtype')
+    return fields
+
+
+def _duplicate_key_id_generator(unit):
+    """duplicate NEVRA unit ID generator
+
+    :param unit: The unit whose NEVRA should be removed
+    :type unit: ContentUnit
+
+    find all duplicate NEVRAs, regardless of repository, for a given content unit type
+
+    This is a memory-efficient pre-filter, which finds all possible duplicate NEVRAs in the
+    collection of a content unit that supports the NEVRA fields. This iterator can then be
+    cross-referenced with a repository to remove duplicate repository content unit
+    links for a given NEVRA.
+    """
+    fields = _duplicate_key_nevra_fields(unit)
+
+    try:
+        # this test aggregation is needed because the aggregation generator body
+        # won't raise OperationFailure until its first iteration, at which point it's too late to
+        # switch to the mapreduce generator. this also ensures that if OperationFailure is raised
+        # at this point, it's specifically because an unsupported aggregation feature was requested
+        unit.objects.aggregate({'$limit': 1})
+        return _duplicate_key_id_generator_aggregation(unit, fields)
+    except OperationFailure:
+        # mongodb doesn't support aggregation cursors, use the slower mapreduce method
+        _logger.info('Purging duplicate NEVRA can take significantly longer in versions of '
+                     'mongodb lower than 2.6. Consider upgrading mongodb if cleaning duplicate '
+                     'packages takes an unreasonably long time.')
```

Comment:
We usually gettext our human readable messages, despite not having any translation files.